File: backend/app/schemas/tournament_structure.py
"""Pydantic v2 schemas for Dynamic Tournament Structure feature."""
from __future__ import annotations

from datetime import datetime
from typing import Optional, List, Literal, Any

from pydantic import BaseModel, Field


# ── Node schemas ──────────────────────────────────────────────────────────────

class NodeResponse(BaseModel):
    id:            int
    tournament_id: int
    parent_id:     Optional[int]
    level:         int
    node_type:     Literal["group", "weight_class"]
    name:          str
    node_code:     Optional[str] = None   # machine-readable code, e.g. "M","F","1A","45"
    rule_json:     Optional[Any] = None   # eligibility rules: min_age, max_age, allowed_belts, max_weight_kg
    sort_order:    int
    student_count: int
    created_at:    datetime
    updated_at:    datetime

    model_config = {"from_attributes": True}


class NodeTreeResponse(NodeResponse):
    """NodeResponse with recursive children list (only used in tree format)."""
    children: List[NodeTreeResponse] = []


# Pydantic v2 requires update_forward_refs call style via model_rebuild
NodeTreeResponse.model_rebuild()


class GetNodesResponse(BaseModel):
    tournament_id:      int
    tournament_name:    str
    tournament_status:  str
    nodes:              list  # NodeTreeResponse[] or NodeResponse[] depending on format
    stats: dict  # total_nodes, total_weight_classes, total_students


class CreateNodeRequest(BaseModel):
    parent_id: Optional[int] = None
    name:      str = Field(..., max_length=100)
    node_type: Optional[Literal["group", "weight_class"]] = None
    node_code: Optional[str] = Field(None, max_length=50,
        description="Machine-readable code (e.g. 'M','F','1A','45'). Không parse name.")
    rule_json: Optional[Any] = Field(None,
        description="Eligibility rules: {min_age, max_age, allowed_belts, max_weight_kg, ...}")


class UpdateNodeRequest(BaseModel):
    name: str = Field(..., max_length=100)


class DeleteNodeRequest(BaseModel):
    move_to_node_id: Optional[int] = None


class DeleteNodeResponse(BaseModel):
    deleted_node_id: int
    deleted_count:   int
    moved_students:  int


class ReorderNodeItem(BaseModel):
    node_id:    int
    sort_order: int


class ReorderNodesRequest(BaseModel):
    parent_id: Optional[int] = None
    nodes:     List[ReorderNodeItem]


class ReorderNodesResponse(BaseModel):
    updated_count: int
    nodes:         List[NodeResponse]


class CopyStructureRequest(BaseModel):
    source_tournament_id: int
    copy_katas:           bool = True


class CopyStructureResponse(BaseModel):
    copied_nodes: int
    copied_katas: int
    tree:         List[NodeTreeResponse]


# ── Node students schemas ─────────────────────────────────────────────────────

class StudentInNodeItem(BaseModel):
    student_id:         int
    student_name:       str
    assigned_node_id:   int
    assigned_node_name: str
    contest_types:      List["ContestTypeItem"]


class NodeStudentsResponse(BaseModel):
    node_id:   int
    node_path: str
    students:  List[StudentInNodeItem]
    total:     int


# ── Kata schemas ──────────────────────────────────────────────────────────────

class KataResponse(BaseModel):
    id:            int
    tournament_id: int
    division:      Literal["individual", "team"]
    name:          str
    description:   Optional[str]
    sort_order:    int
    usage_count:   int
    team_size:     int = 2
    min_team_size: Optional[int] = None

    model_config = {"from_attributes": True}


class ListKatasResponse(BaseModel):
    tournament_id: int
    katas:         List[KataResponse]
    total:         int


class CreateKataRequest(BaseModel):
    division:     Literal["individual", "team"] = "individual"
    name:         str = Field(..., max_length=100)
    description:  Optional[str] = Field(None, max_length=500)
    team_size:    int = Field(2, ge=2)
    min_team_size: Optional[int] = Field(None, ge=2)


class UpdateKataRequest(BaseModel):
    division:     Optional[Literal["individual", "team"]] = None
    name:         Optional[str] = Field(None, max_length=100)
    description:  Optional[str] = Field(None, max_length=500)
    team_size:    Optional[int] = Field(None, ge=2)
    min_team_size: Optional[int] = Field(None, ge=2)


class ReorderKataItem(BaseModel):
    kata_id:    int
    sort_order: int


class ReorderKatasRequest(BaseModel):
    katas: List[ReorderKataItem]


class ReorderKatasResponse(BaseModel):
    updated_count: int
    katas:         List[KataResponse]


class DeleteKataResponse(BaseModel):
    deleted_kata_id: int
    kata_name:       str


class TeamKataRegistrationItem(BaseModel):
    node_id: int
    node_path: str
    kata_id: int
    kata_name: str


class TeamKataRegistrationResponse(BaseModel):
    tournament_id: int
    club_id: int
    registrations: List[TeamKataRegistrationItem]


class TeamKataRegistrationSelection(BaseModel):
    node_id: int
    kata_id: int


class UpdateTeamKataRegistrationRequest(BaseModel):
    items: List[TeamKataRegistrationSelection] = Field(default_factory=list)


class TeamKataMemberItem(BaseModel):
    student_id:   int
    student_name: str
    club_name:    str


class TeamKataMembersResponse(BaseModel):
    tournament_id: int
    club_id:       int
    node_id:       int
    kata_id:       int
    team_size:     int
    members:       List[TeamKataMemberItem]


class UpdateTeamKataMembersRequest(BaseModel):
    student_ids: List[int] = Field(default_factory=list)


# ── Contest type / registration schemas ──────────────────────────────────────

class ContestTypeItem(BaseModel):
    type:      Literal["sparring", "kata"]
    kata_id:   Optional[int] = None
    kata_name: Optional[str] = None


class RegisterParticipantRequest(BaseModel):
    node_id: int  # Required: parent of weight_class (classification node)
    sparring: bool
    sparring_weight_id: Optional[int] = None  # Required if sparring=true (weight_class leaf)
    kata: bool
    kata_ids: List[int] = Field(default_factory=list)  # Required if kata=true (min_length=1)


class UpdateContestTypesRequest(BaseModel):
    contest_types: List[ContestTypeItem] = Field(..., min_length=1)


class ReassignNodeRequest(BaseModel):
    new_node_id: int


class StudentRegistrationResponse(BaseModel):
    student_id:    int
    tournament_id: int
    node_id:       int
    classification_node_id: Optional[int] = None
    sparring_weight_id: Optional[int] = None
    node_path:     str
    registered_at: datetime
    contest_types: List[ContestTypeItem]


# ── Eligible nodes schemas ────────────────────────────────────────────────────

class EligibleNodeItem(BaseModel):
    node_id:           int
    path:              str
    node_code:         Optional[str]
    reason:            str   # "recommended" | "override_allowed"
    eligible:          bool
    ineligible_reason: Optional[str] = None


class EligibleNodesResponse(BaseModel):
    tournament_id:      int
    student_id:         int
    recommended_node_id: Optional[int]
    recommended_path:   Optional[str]
    candidate_nodes:    List[EligibleNodeItem]
    warnings:           List[str] = []


# ── Atomic register-student schemas ──────────────────────────────────────────

class StudentCreateInline(BaseModel):
    """Student fields for atomic create+register."""
    full_name      : str
    gender         : str
    club_id        : int
    current_belt   : str = "Tự vệ nhập môn"
    phone          : Optional[str] = None
    email          : Optional[str] = None
    address        : Optional[str] = None
    date_of_birth  : Optional[Any] = None
    id_number      : Optional[str] = None
    join_date      : Optional[Any] = None
    belt_date      : Optional[Any] = None
    competition_weight_kg: Optional[float] = None
    notes          : Optional[str] = None


class RegisterStudentRequest(BaseModel):
    """Atomic: tạo student mới + đăng ký vào dynamic tournament trong 1 transaction."""
    student:        StudentCreateInline
    node_id:        int
    contest_types:  List[ContestTypeItem] = Field(..., min_length=1)
    override_reason: Optional[str] = None  # bắt buộc khi đăng ký ngoài node recommended


class RegisterStudentResponse(BaseModel):
    student_id:   int
    student_code: str
    node_path:    str
    registration: StudentRegistrationResponse


# ── Bracket tree schemas ──────────────────────────────────────────────────────

class BracketLeafParticipant(BaseModel):
    student_id: int
    full_name: str


class BracketLeafInfo(BaseModel):
    weight_class_id: int
    bracket_status: str    # NOT_GENERATED | GENERATING | GENERATED
    total_players: int
    players: list[str]
    participants: list[BracketLeafParticipant] = []


class BracketNodeItem(BaseModel):
    id: int
    name: str
    node_type: Literal["group", "weight_class"]
    sort_order: int
    leaf_info: Optional[BracketLeafInfo] = None  # populated on weight_class nodes only
    children: list["BracketNodeItem"] = []


BracketNodeItem.model_rebuild()


class BracketTreeResponse(BaseModel):
    tournament_id: int
    tournament_name: str
    tournament_status: str
    nodes: list[BracketNodeItem]

File: frontend/src/api/tournament_structure.ts
import api from '../lib/axios'
import type {
  TournamentStructureNode,
  TournamentKata,
  ContestTypeItem,
  StudentRegistration,
  NodesResponse,
  KatasResponse,
  TeamKataRegistrationResponse,
  TeamKataMembersResponse,
  NodeStudentsResponse,
  BracketTreeResponse,
} from '../types/tournament'

// ── Node endpoints ─────────────────────────────────────────────────────────

export const getNodes = async (
  tournamentId: number,
  format: 'tree' | 'flat' = 'flat',
): Promise<NodesResponse> => {
  const res = await api.get<NodesResponse>(
    `/api/tournaments/${tournamentId}/weight-class-nodes`,
    { params: { format } },
  )
  return res.data
}

export interface CreateNodePayload {
  parent_id: number | null
  name:      string
  node_type?: 'group' | 'weight_class'
}

export const createNode = async (
  tournamentId: number,
  data: CreateNodePayload,
): Promise<TournamentStructureNode> => {
  const res = await api.post<TournamentStructureNode>(
    `/api/tournaments/${tournamentId}/weight-class-nodes`,
    data,
  )
  return res.data
}

export const updateNode = async (
  tournamentId: number,
  nodeId: number,
  data: { name: string },
): Promise<TournamentStructureNode> => {
  const res = await api.patch<TournamentStructureNode>(
    `/api/tournaments/${tournamentId}/weight-class-nodes/${nodeId}`,
    data,
  )
  return res.data
}

export interface DeleteNodePayload {
  move_to_node_id?: number | null
}

export interface DeleteNodeResult {
  deleted_node_id: number
  deleted_count:   number
  moved_students:  number
}

export const deleteNode = async (
  tournamentId: number,
  nodeId: number,
  data: DeleteNodePayload,
): Promise<DeleteNodeResult> => {
  const res = await api.delete<DeleteNodeResult>(
    `/api/tournaments/${tournamentId}/weight-class-nodes/${nodeId}`,
    { data },
  )
  return res.data
}

export interface ReorderNodeItem {
  node_id:    number
  sort_order: number
}

export interface ReorderNodesPayload {
  parent_id: number | null
  nodes:     ReorderNodeItem[]
}

export const reorderNodes = async (
  tournamentId: number,
  data: ReorderNodesPayload,
): Promise<{ updated_count: number; nodes: TournamentStructureNode[] }> => {
  const res = await api.post(
    `/api/tournaments/${tournamentId}/weight-class-nodes/reorder`,
    data,
  )
  return res.data
}

export interface CopyStructurePayload {
  source_tournament_id: number
  copy_katas:           boolean
}

export interface CopyStructureResult {
  copied_nodes: number
  copied_katas: number
  tree:         TournamentStructureNode[]
}

export const copyStructure = async (
  tournamentId: number,
  data: CopyStructurePayload,
): Promise<CopyStructureResult> => {
  const res = await api.post<CopyStructureResult>(
    `/api/tournaments/${tournamentId}/weight-class-nodes/copy`,
    data,
  )
  return res.data
}

export const getNodeStudents = async (
  tournamentId: number,
  nodeId: number,
  includeDescendants = true,
): Promise<NodeStudentsResponse> => {
  const res = await api.get<NodeStudentsResponse>(
    `/api/tournaments/${tournamentId}/weight-class-nodes/${nodeId}/students`,
    { params: { include_descendants: includeDescendants } },
  )
  return res.data
}

// ── Participant endpoints ──────────────────────────────────────────────────

export interface RegisterParticipantPayload {
  node_id: number | null
  sparring: boolean
  sparring_weight_id: number | null
  kata: boolean
  kata_ids: number[]
}

export const registerParticipant = async (
  tournamentId: number,
  studentId: number,
  data: RegisterParticipantPayload,
): Promise<StudentRegistration> => {
  const res = await api.post<StudentRegistration>(
    `/api/tournaments/${tournamentId}/participants/${studentId}/register`,
    data,
  )
  return res.data
}

export const reassignNode = async (
  tournamentId: number,
  studentId: number,
  data: { new_node_id: number },
): Promise<StudentRegistration> => {
  const res = await api.post<StudentRegistration>(
    `/api/tournaments/${tournamentId}/participants/${studentId}/reassign`,
    data,
  )
  return res.data
}

export interface UpdateContestTypesPayload {
  contest_types: ContestTypeItem[]
}

export const updateContestTypes = async (
  tournamentId: number,
  studentId: number,
  data: UpdateContestTypesPayload,
): Promise<StudentRegistration> => {
  const res = await api.patch<StudentRegistration>(
    `/api/tournaments/${tournamentId}/participants/${studentId}/contest-types`,
    data,
  )
  return res.data
}

export const getParticipantRegistration = async (
  tournamentId: number,
  studentId: number,
): Promise<StudentRegistration> => {
  const res = await api.get<StudentRegistration>(
    `/api/tournaments/${tournamentId}/participants/${studentId}`,
  )
  return res.data
}

// ── Kata endpoints ─────────────────────────────────────────────────────────

export const listKatas = async (tournamentId: number): Promise<KatasResponse> => {
  const res = await api.get<KatasResponse>(`/api/tournaments/${tournamentId}/katas`)
  return res.data
}

export interface CreateKataPayload {
  division:    'individual' | 'team'
  name:        string
  description?: string | null
  team_size?:  number
}

export const createKata = async (
  tournamentId: number,
  data: CreateKataPayload,
): Promise<TournamentKata> => {
  const res = await api.post<TournamentKata>(`/api/tournaments/${tournamentId}/katas`, data)
  return res.data
}

export interface UpdateKataPayload {
  division?:     'individual' | 'team'
  name?:         string
  description?:  string | null
  team_size?:    number | null
  min_team_size?: number | null
}

export const updateKata = async (
  tournamentId: number,
  kataId: number,
  data: UpdateKataPayload,
): Promise<TournamentKata> => {
  const res = await api.patch<TournamentKata>(
    `/api/tournaments/${tournamentId}/katas/${kataId}`,
    data,
  )
  return res.data
}

export interface DeleteKataResult {
  deleted_kata_id: number
  kata_name:       string
}

export const deleteKata = async (
  tournamentId: number,
  kataId: number,
): Promise<DeleteKataResult> => {
  const res = await api.delete<DeleteKataResult>(
    `/api/tournaments/${tournamentId}/katas/${kataId}`,
  )
  return res.data
}

export interface ReorderKataItem {
  kata_id:    number
  sort_order: number
}

export const reorderKatas = async (
  tournamentId: number,
  katas: ReorderKataItem[],
): Promise<{ updated_count: number; katas: TournamentKata[] }> => {
  const res = await api.post(`/api/tournaments/${tournamentId}/katas/reorder`, { katas })
  return res.data
}

// ── Bracket tree (dynamic tournament bracket display) ────────────────────────

export const getBracketTree = async (tournamentId: number): Promise<BracketTreeResponse> => {
  const res = await api.get<BracketTreeResponse>(`/api/tournaments/${tournamentId}/bracket-tree`)
  return res.data
}

export const getTeamKataRegistrations = async (
  tournamentId: number,
  clubId: number,
): Promise<TeamKataRegistrationResponse> => {
  const res = await api.get<TeamKataRegistrationResponse>(
    `/api/tournaments/${tournamentId}/clubs/${clubId}/team-kata-registrations`,
  )
  return res.data
}

export const replaceTeamKataRegistrations = async (
  tournamentId: number,
  clubId: number,
  items: { node_id: number; kata_id: number }[],
): Promise<TeamKataRegistrationResponse> => {
  const res = await api.put<TeamKataRegistrationResponse>(
    `/api/tournaments/${tournamentId}/clubs/${clubId}/team-kata-registrations`,
    { items },
  )
  return res.data
}

export const getTeamKataMembers = async (
  tournamentId: number,
  clubId: number,
  nodeId: number,
  kataId: number,
): Promise<TeamKataMembersResponse> => {
  const res = await api.get<TeamKataMembersResponse>(
    `/api/tournaments/${tournamentId}/clubs/${clubId}/team-kata-members`,
    { params: { node_id: nodeId, kata_id: kataId } },
  )
  return res.data
}

export const replaceTeamKataMembers = async (
  tournamentId: number,
  clubId: number,
  nodeId: number,
  kataId: number,
  studentIds: number[],
): Promise<TeamKataMembersResponse> => {
  const res = await api.put<TeamKataMembersResponse>(
    `/api/tournaments/${tournamentId}/clubs/${clubId}/team-kata-members`,
    { student_ids: studentIds },
    { params: { node_id: nodeId, kata_id: kataId } },
  )
  return res.data
}
